internal/argocd/health.go

```diff
@@ -217,58 +224,123 @@ func (h *applicationHealth) GetApplicationHealth(
 		//
 		// TODO: revisit this when https://github.com/argoproj/argo-cd/pull/18660
 		// 	 is merged and released.
-		cooldown := app.Status.OperationState.FinishedAt.Time.Add(10 * time.Second)
-		if duration := time.Until(cooldown); duration > 0 {
-			time.Sleep(duration)
-
-			// Re-fetch the application to get the latest state.
-			if err := h.argoClient.Get(ctx, key, app); err != nil {
-				err = fmt.Errorf("error finding Argo CD Application %q in namespace %q: %w", key.Name, key.Namespace, err)
-				if client.IgnoreNotFound(err) == nil {
-					err = fmt.Errorf("unable to find Argo CD Application %q in namespace %q", key.Name, key.Namespace)
+		if app.Status.OperationState != nil {
```

Comment:
This new nil-check looks like it probably should have been there all along, however, when you added it, you also added the possibility that one or more desired revisions are known, but we still fall through to the logic at the very end of the function that deems the Application healthy without further conditions.

Reply:
> but we still fall through to the logic at the very end of the function that deems the Application healthy without further conditions.

Sorry. I read something wrong. This isn't the case.

@hiddeco knows this part of the code better. What's the right thing to do if we don't have any operation state?

Reply:
`stageHealthForAppSync` used to error out if there was no `OperationState`, this is because ArgoCD resets the `OperationState` when a new `Operation` is requested (https://github.com/argoproj/argo-cd/blob/473665795c9bcc40e3a60e3bddb7edf76747c974/util/argo/argo.go#L800-L801).

Given this, the lack of an `OperationState` can indicate that a new (sync) operation is being attempted (or has never happened).

Reply:
So, keep, remove, what is expected?